evm/src/lib.rs

```diff
@@ -1,3 +1,153 @@
+//! );
+//! ```
+//!
+//! # Inputs type
+//!
+//! Transactions need to be processed into an Intermediary Representation (IR) format for the prover
+//! to be able to generate proofs of valid state transition. This involves passing the encoded transaction,
+//! the header of the block in which it was included, some information on the state prior execution
+//! of this transaction, etc.
+//! This intermediary representation is called [`GenerationInputs`].
+//!
+//!
+//! # Generating succinct proofs
+//!
+//! ## Transaction proofs
+//!
+//! To generate a proof for a transaction, given its [`GenerationInputs`] and an [`AllRecursiveCircuits`]
+//! prover state, one can simply call the [prove_root](AllRecursiveCircuits::prove_root) method.
+//!
+//! ```ignore
+//! let mut timing = TimingTree::new("prove", log::Level::Debug);
+//! let kill_signal = None; // Useful only with distributed proving to kill hanging jobs.
+//! let (proof, public_values) =
+//!     prover_state.prove_root(all_stark, config, inputs, &mut timing, kill_signal);
+//! ```
+//!
+//! This outputs a transaction proof and its associated public values. These are necessary during the
+//! aggregation levels (see below). If one were to miss the public values, they are also retrievable directly
+//! from the proof's encoded public inputs, as such:
+//!
+//! ```ignore
+//! let public_values = PublicValues::from_public_inputs(&proof.public_inputs);
+//! ```
+//!
+//! ## Aggregation proofs
+//!
+//! Because the plonky2 zkEVM generates proofs on a transaction basis, we then need to aggregate them for succinct
+//! verification. This is done in a binary tree fashion, where each inner node proof verifies two children proofs,
+//! through the [prove_aggreg](AllRecursiveCircuits::prove_aggreg) method.
+//! Note that the tree does *not* need to be complete, as this aggregation process can take as inputs both regular
+//! transaction proofs and aggregation proofs. We only need to specify for each child if it is an aggregation proof
+//! or a regular one.
+//!
+//! ```ignore
+//! let (proof_1, pv_1) =
+//!     prover_state.prove_root(all_stark, config, inputs_1, &mut timing, None);
+//! let (proof_2, pv_2) =
+//!     prover_state.prove_root(all_stark, config, inputs_2, &mut timing, None);
+//! let (proof_3, pv_3) =
+//!     prover_state.prove_root(all_stark, config, inputs_2, &mut timing, None);
+//!
+//! // Now aggregate proofs for txn 1 and 2.
+//! let (agg_proof_1_2, pv_1_2) =
+//!     prover_state.prove_aggreg(false, proof_1, pv_1, false, proof_2, pv_2);
+//!
+//! // Now aggregate the newly generated aggregation proof with the last regular txn proof.
+//! let (agg_proof_1_3, pv_1_3) =
+//!     prover_state.prove_aggreg(true, agg_proof_1_2, pv_1_2, false, proof_3, pv_3);
+//! ```
+//!
+//! **Note**: The proofs provided to the [prove_aggreg](AllRecursiveCircuits::prove_aggreg) method *MUST* have contiguous states.
+//! Trying to combine `proof_1` and `proof_3` from the example above would fail.
```

Comment:
This was a question I had: are we guaranteed to fail if/when states are not contiguous (but otherwise valid)?

Reply:
Yes, because we try connecting the states (with some equality checks in the circuit), for instance that the gas used after txn 1 is equal to the gas used prior txn 2. So the prover would not be able to generate a proof.